### PR TITLE
Handle init_blocks in scaling strategy, rather than special-casing it

### PR DESCRIPTION
**Pull request by benclifford**
This is part of issue #3278 tidying up job and block management.

# Changed Behaviour

Now init_blocks scale out happens on the first strategy poll, not at executor start - that will often delay init_blocks scaling by one strategy poll period compared to before this PR.

## Type of change

- Code maintenance/cleanup

### CODE REVIEW

**Comment by benclifford** [comment]
failing a precondition in test_drain for exactly the "Changed Behaviour" reason - now that test will be reliant on strategy poll period so it can probably be made to use a smaller strategy period

File: parsl/jobs/strategy.py [hunk 1]
```diff
@@ -146,10 +146,17 @@ def add_executors(self, executors: Sequence[ParslExecutor]) -> None:
...
         """
-        logger.debug("strategy_noop: doing nothing")
+        for exec_status in status_list:
+            if exec_status.first:
+                executor = exec_status.executor
+                logger.debug(f"strategy_init_only: scaling out {executor.provider.init_blocks} initial blocks for {executor.label}")
+                exec_status.scale_out(executor.provider.init_blocks)
+                exec_status.first = False
+            else:
+                logger.debug("strategy_init_only: doing nothing")
```
Comment (khk-globus) [inline parsl/jobs/strategy.py:159]: Is this prefix necessary?  I thought we included the method name in the log line already, per the log formatter.

File: parsl/jobs/strategy.py [hunk 2]
```diff
@@ -146,10 +146,17 @@ def add_executors(self, executors: Sequence[ParslExecutor]) -> None:
...
-        """Do nothing.
+    def _strategy_init_only(self, status_list: List[jsp.PollItem]) -> None:
+        """Scale up to init_blocks at the start, then nothing more.
         """
-        logger.debug("strategy_noop: doing nothing")
+        for exec_status in status_list:
+            if exec_status.first:
+                executor = exec_status.executor
+                logger.debug(f"strategy_init_only: scaling out {executor.provider.init_blocks} initial blocks for {executor.label}")
+                exec_status.scale_out(executor.provider.init_blocks)
```
Comment (khk-globus) [inline parsl/jobs/strategy.py:156]: For new SLOCs, might consider passing the args to `.debug()` rather that doing the formatting up front.  No sense in doing the work if debug-level logs aren't emitted:
```python
logger.debug("scaling out %s initial blocks for %s", executor.provider.init_blocks, executor.label)
```

**Review by khk-globus — APPROVED** [review]
Looks good, with a couple of minor comments.  A good addition to this would be a test verifying use of the `.first` branch.

File: parsl/jobs/strategy.py (hunk 1 above)
Comment (benclifford) [reply parsl/jobs/strategy.py:159]: It's a vestige from the line it used to be, which included that (renamed) prefix.

The method name is indeed in the log formatter

File: parsl/jobs/strategy.py (hunk 2 above)
Comment (benclifford) [reply parsl/jobs/strategy.py:156]: this isn't a new line - it's a move and a keyword rename!

**Comment by benclifford** [comment]
> A good addition to this would be a test verifying use of the .first branch

From an end functionality perspective, that's checking that init_blocks actually gets used to start blocks, rather than a later part of the strategy code: with strategy `none` that's pretty straightforward (and it's what happens by chance in parsl/tests/test_monitoring/test_htex_init_blocks_vs_monitoring.py already).

With the other two strategies that's less observable from the outside, I think: `init_blocks` vs `min_blocks` doesn't necessarily even make sense because of that difficulty in observing the difference...

File: parsl/jobs/strategy.py (hunk 2 above)
Comment (khk-globus) [reply parsl/jobs/strategy.py:156]: :boom:

**Pull request merged**